### PR TITLE
[HDF5][BugFix] Support for recursive submodel parts.

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This fixes a bug in the recursive submodel part handling. Thanks for the small test case @philbucher. It helped a lot :)

**Changelog**
- Fixes recursive submodel part handling for output

### CODE REVIEW

**Comment by philbucher** [comment]
wow this is AMAZING!
Using the Multi-block Inspector one can view the Model in so much detail! This is so much better than what we can do with vtk!!!

Please add a test then I will approve

@KratosMultiphysics/structural-mechanics this is really cool, I recommend checking this out the HDF output, esp for complicated models!
For me this is definitely my new default output, this combines all the good things (smallest files sizes, writing mesh only once, being able to view the model in so much detail and one can also read the files in python in case one needs to get some values later)

**Comment by sunethwarna** [comment]
@philbucher The test is done :)

File: applications/HDF5Application/custom_io/hdf5_model_part_io.cpp [hunk 1]
```diff
@@ -262,16 +265,22 @@ std::vector<std::size_t> ModelPartIO::ReadContainerIds(std::string const& rPath)
...
-    mpFile->AddPath(GroupName);
-    for (ModelPart const& r_sub_model_part : rModelPart.SubModelParts())
-    {
-        for (ModelPart const& r_sub_sub_model_part : r_sub_model_part.SubModelParts())
-        {
-            WriteSubModelParts(r_sub_sub_model_part, GroupName + "/" + r_sub_model_part.Name());
-        }
+    unsigned start_index, block_size;
+    std::tie(start_index, block_size) = StartIndexAndBlockSize(rPath);
+    Vector<int> id_buf;
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_model_part_io.cpp:272]: cannot you pass an `std::vector` directly?

File: applications/HDF5Application/custom_io/hdf5_model_part_io.cpp [hunk 2]
```diff
@@ -262,16 +265,22 @@ std::vector<std::size_t> ModelPartIO::ReadContainerIds(std::string const& rPath)
...
+    mpFile->ReadDataSet(rPath + "/Ids", id_buf, start_index, block_size);
+    std::vector<std::size_t> ids(id_buf.size());
+#pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(ids.size()); ++i)
+        ids[i] = id_buf[i];
+    return ids;
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_model_part_io.cpp:278]: why not parallel utilities?

**Review by philbucher — COMMENTED** [review]
minor comments otherwise looks good

**Review by philbucher — APPROVED** [review]

**Pull request merged**